bootstrap/src/lib/cache/rateLimit.ts

```diff
@@ -0,0 +1,140 @@
+import { logger } from '@chainlink/external-adapter'
+import { DEFAULT_CACHE_MAX_AGE } from './redis'
+
+const MAX_AGE_ALLOWED = 1000 * 60 * 2
+
+type RateLimitOptions = {
+  groupMaxAge: number
+  groupId: string
+  id: string
+  totalCapacity: number
+}
+
+export interface RateLimit {
+  isEnabled: () => boolean
+  getParticipantMaxAge: (participantId: string) => number | false
+  incrementParticipantHeartbeat: (participantId: string, cost?: number) => number | false
+}
+
+type Participant = {
+  cost: number
+  heartbeat: number
+  lastSeen: number
+}
+
+const totalHeartbeat = new Map()
+let participants: Map<string, Participant> = new Map()
+
+export const makeRateLimit = (options: RateLimitOptions): RateLimit => {
+  const minHeartbeat = 0
+  const safeCapacity = options.totalCapacity * 0.9
+
+  const _isEnabled = (): boolean => {
+    return !!options.groupId && !!options.totalCapacity
+  }
+
+  // Weight normalizes the relevance the participant is having in the adapter
+  const _getWeight = (participantHeartbeat: number, totalHeartbeat: number): number => {
+    // Avoid having Infinity weight
+    if (participantHeartbeat === 0) {
+      participantHeartbeat = 1
+    }
+    if (totalHeartbeat === 0) {
+      totalHeartbeat = 1
+    }
+    return (participantHeartbeat - minHeartbeat) / (totalHeartbeat - minHeartbeat)
+  }
+
+  const _getParticipant = (participantId: string): Participant => {
+    return (
+      participants.get(participantId) || { cost: 1, heartbeat: 0, lastSeen: new Date().getTime() }
+    )
+  }
+
+  const _getParticipantEvenCapacity = (participantId: string) => {
+    const totalParticipants = participants.size || 1
+    return safeCapacity / totalParticipants / _getParticipant(participantId).cost
+  }
+
+  const _getParticipantMaxAge = (participantId: string): number => {
+    const SEC_IN_MIN = 60
+    const MS_IN_SEC = 1000
+
+    // Precise Capacity
+    // const participantHeartbeat = _getParticipant(participantId).heartbeat
+    // const totalHeartbeat = _getCurrentHeartbeat()
+    // const participantWeight = _getWeight(participantHeartbeat, totalHeartbeat)
+    // const allowedReqPerMin = safeCapacity * participantWeight
+
+    const allowedReqPerMin = _getParticipantEvenCapacity(participantId)
+    console.log('PARTICIPANT CAPACITY: ', allowedReqPerMin)
+
+    let maxAge = Math.round(MS_IN_SEC / (allowedReqPerMin / SEC_IN_MIN))
```

Comment:
I would leave a quick comment here just to make it easier for future readers to comprehend:
`// the max age we set becomes the average age of a request coming in`
`// e.g. if we have 5 req per min, then the max age will be 250ms, since that means on average we have a request coming in every 250ms`
`// the calculated max age is bounded to the range of [DEFAULT_CACHE_MAX_AGE, MAX_AGE_ALLOWED] `